How to mock and test request with content passed as byte iterator

Hello team, 

I have `httpx` code that passes `content` to request using byte iterator. When trying the test code using `pytest_httpx` and `match_content` on mocked request, I get assertion error and the `aiter_bytes` is never called. Here's what I'm trying...

```python
import httpx
import pytest


async def aiter_bytes():
    yield b"h"
    yield b"e"
    yield b"l"
    yield b"l"
    yield b"o"


@pytest.mark.asyncio
async def test_httpx_get(httpx_mock):
    httpx_mock.add_response(method="PUT", url="http://put", content=b"world", match_content=b"hello")
    async with httpx.AsyncClient() as client:
        res = await client.put("http://put", content=aiter_bytes())
        assert res.text == "world"
```

and the exception when runnning the test
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Request('PUT', 'http://put')>

    def read(self) -> bytes:
        """
        Read and return the request content.
        """
        if not hasattr(self, "_content"):
>           assert isinstance(self.stream, typing.Iterable)
E           AssertionError
```

When I remove the `match_content` parameter, the test passes, but the `aiter_bytes` is never called. Is there anything I'm missing? I'm not sure whether this is `pytest_httpx` or `httpx` issue, although when I run the code without `pytest` in real program, it works and sends the data properly.

Thank you,
Adam

> Hello,
> 
> [The documentation is already explaining how to stream chunks](https://github.com/Colin-b/pytest_httpx?tab=readme-ov-file#reply-by-streaming-chunks)
> 
> If something is not working in the provided example, feel free to report it.
> 
> Thanks

Hi @Colin-b, thanks for reaching out. I'am aware of the documentation, although the docs covers the case when mocked url produce the stream, which is not wham I'm trying to achieve.

Following code passes fine
```
async def test_httpx_put(httpx_mock):
    httpx_mock.add_response(method="PUT", url="http://put", content=b"world", match_content=b"hello")
    async with httpx.AsyncClient() as client:
        res = await client.put("http://put", content=b"hello")
        assert res.text == "world"
```
and checks that I call http://put with "hello" content.

when I change the implementation of sending the data to async iterator as follows
```python
async def aiter_bytes():
    yield b"h"
    yield b"e"
    yield b"l"
    yield b"l"
    yield b"o"


@pytest.mark.asyncio
async def test_httpx_put_iter(httpx_mock):
    httpx_mock.add_response(method="PUT", url="http://put", content=b"world", match_content=b"hello")
    async with httpx.AsyncClient() as client:
        res = await client.put("http://put", content=aiter_bytes())
        assert res.text == "world"
```
the execution fails with error 
```
self = <Request('PUT', 'http://put')>

    def read(self) -> bytes:
        """
        Read and return the request content.
        """
        if not hasattr(self, "_content"):
>           assert isinstance(self.stream, typing.Iterable)
E           AssertionError
```

The code should work, based on httpx documentation you can pass content as _Binary content to include in the body of the request, as bytes or a byte iterator._

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.31.1] - 2024-09-22
+### Fixed
+- It is now possible to match on content provided as async iterable by the client.
+
 ## [0.31.0] - 2024-09-20
 ### Changed
 - Tests will now fail at teardown by default if some requests were issued but were not matched.
@@ -337,7 +341,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - First release, should be considered as unstable for now as design might change.
 
-[Unreleased]: https://github.com/Colin-b/pytest_httpx/compare/v0.31.0...HEAD
+[Unreleased]: https://github.com/Colin-b/pytest_httpx/compare/v0.31.1...HEAD
+[0.31.1]: https://github.com/Colin-b/pytest_httpx/compare/v0.31.0...v0.31.1
 [0.31.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.30.0...v0.31.0
 [0.30.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.29.0...v0.30.0
 [0.29.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.28.0...v0.29.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+
+## [0.31.0] - 2024-09-20
 ### Changed
 - Tests will now fail at teardown by default if some requests were issued but were not matched.
   - This behavior can be changed thanks to the new ``pytest.mark.httpx_mock(assert_all_requests_were_expected=False)`` option.
@@ -335,7 +337,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - First release, should be considered as unstable for now as design might change.
 
-[Unreleased]: https://github.com/Colin-b/pytest_httpx/compare/v0.30.0...HEAD
+[Unreleased]: https://github.com/Colin-b/pytest_httpx/compare/v0.31.0...HEAD
+[0.31.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.30.0...v0.31.0
 [0.30.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.29.0...v0.30.0
 [0.29.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.28.0...v0.29.0
 [0.28.0]: https://github.com/Colin-b/pytest_httpx/compare/v0.27.0...v0.28.0

--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 <a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Build status" src="https://github.com/Colin-b/pytest_httpx/workflows/Release/badge.svg"></a>
 <a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Coverage" src="https://img.shields.io/badge/coverage-100%25-brightgreen"></a>
 <a href="https://github.com/psf/black"><img alt="Code style: black" src="https://img.shields.io/badge/code%20style-black-000000.svg"></a>
-<a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Number of tests" src="https://img.shields.io/badge/tests-211 passed-blue"></a>
+<a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Number of tests" src="https://img.shields.io/badge/tests-215 passed-blue"></a>
 <a href="https://pypi.org/project/pytest-httpx/"><img alt="Number of downloads" src="https://img.shields.io/pypi/dm/pytest_httpx"></a>
 </p>
 

--- a/README.md
+++ b/README.md
@@ -5,7 +5,7 @@
 <a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Build status" src="https://github.com/Colin-b/pytest_httpx/workflows/Release/badge.svg"></a>
 <a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Coverage" src="https://img.shields.io/badge/coverage-100%25-brightgreen"></a>
 <a href="https://github.com/psf/black"><img alt="Code style: black" src="https://img.shields.io/badge/code%20style-black-000000.svg"></a>
-<a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Number of tests" src="https://img.shields.io/badge/tests-208 passed-blue"></a>
+<a href="https://github.com/Colin-b/pytest_httpx/actions"><img alt="Number of tests" src="https://img.shields.io/badge/tests-211 passed-blue"></a>
 <a href="https://pypi.org/project/pytest-httpx/"><img alt="Number of downloads" src="https://img.shields.io/pypi/dm/pytest_httpx"></a>
 </p>
 

--- a/pytest_httpx/_httpx_mock.py
+++ b/pytest_httpx/_httpx_mock.py
@@ -1,8 +1,8 @@
 import copy
 import inspect
 from operator import methodcaller
-from typing import Union, Optional, Callable, Any, NoReturn
-from collections.abc import Awaitable
+from typing import Union, Optional, Callable, Any, NoReturn, AsyncIterable
+from collections.abc import Awaitable, Iterable
 
 import httpx
 from pytest import Mark
@@ -157,6 +157,8 @@ class HTTPXMock:
         real_transport: httpx.HTTPTransport,
         request: httpx.Request,
     ) -> httpx.Response:
+        # Store the content in request for future matching
+        request.read()
         self._requests.append((real_transport, request))
 
         callback = self._get_callback(real_transport, request)
@@ -173,6 +175,8 @@ class HTTPXMock:
         real_transport: httpx.AsyncHTTPTransport,
         request: httpx.Request,
     ) -> httpx.Response:
+        # Store the content in request for future matching
+        await request.aread()
         self._requests.append((real_transport, request))
 
         callback = self._get_callback(real_transport, request)

--- a/pytest_httpx/_request_matcher.py
+++ b/pytest_httpx/_request_matcher.py
@@ -98,11 +98,12 @@ class _RequestMatcher:
     def _content_match(self, request: httpx.Request) -> bool:
         if self.content is None and self.json is None:
             return True
+
         if self.content is not None:
-            return request.read() == self.content
+            return request.content == self.content
         try:
             # httpx._content.encode_json hard codes utf-8 encoding.
-            return json.loads(request.read().decode("utf-8")) == self.json
+            return json.loads(request.content.decode("utf-8")) == self.json
         except json.decoder.JSONDecodeError:
             return False
 

--- a/pytest_httpx/version.py
+++ b/pytest_httpx/version.py
@@ -3,4 +3,4 @@
 # Major should be incremented in case there is a breaking change. (eg: 2.5.8 -> 3.0.0)
 # Minor should be incremented in case there is an enhancement. (eg: 2.5.8 -> 2.6.0)
 # Patch should be incremented in case there is a bug fix. (eg: 2.5.8 -> 2.5.9)
-__version__ = "0.30.0"
+__version__ = "0.31.0"

--- a/pytest_httpx/version.py
+++ b/pytest_httpx/version.py
@@ -3,4 +3,4 @@
 # Major should be incremented in case there is a breaking change. (eg: 2.5.8 -> 3.0.0)
 # Minor should be incremented in case there is an enhancement. (eg: 2.5.8 -> 2.6.0)
 # Patch should be incremented in case there is a bug fix. (eg: 2.5.8 -> 2.5.9)
-__version__ = "0.31.0"
+__version__ = "0.31.1"

--- a/tests/test_httpx_async.py
+++ b/tests/test_httpx_async.py
@@ -2,6 +2,7 @@ import asyncio
 import math
 import re
 import time
+from collections.abc import AsyncIterable
 
 import httpx
 import pytest
@@ -2007,7 +2008,7 @@ async def test_streams_are_not_cascading_resulting_in_maximum_recursion(
 ) -> None:
     httpx_mock.add_response(json={"abc": "def"})
     async with httpx.AsyncClient() as client:
-        tasks = [client.get("https://example.com/") for _ in range(950)]
+        tasks = [client.get("https://test_url") for _ in range(950)]
         await asyncio.gather(*tasks)
     # No need to assert anything, this test case ensure that no error was raised by the gather
 
@@ -2033,3 +2034,61 @@ async def test_custom_transport(httpx_mock: HTTPXMock) -> None:
         response = await client.post("https://test_url", content=b"This is the body")
         assert response.read() == b""
         assert response.headers["x-prefix"] == "test"
+
+
+@pytest.mark.asyncio
+async def test_response_selection_content_matching_with_async_iterable(
+    httpx_mock: HTTPXMock,
+) -> None:
+    httpx_mock.add_response(match_content=b"full content 1", content=b"matched 1")
+    httpx_mock.add_response(match_content=b"full content 2", content=b"matched 2")
+
+    async def stream_content_1() -> AsyncIterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 1"
+
+    async def stream_content_2() -> AsyncIterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 2"
+
+    async with httpx.AsyncClient() as client:
+        response_2 = await client.put("https://test_url", content=stream_content_2())
+        response_1 = await client.put("https://test_url", content=stream_content_1())
+    assert response_1.content == b"matched 1"
+    assert response_2.content == b"matched 2"
+
+
+@pytest.mark.asyncio
+async def test_request_selection_content_matching_with_async_iterable(
+    httpx_mock: HTTPXMock,
+) -> None:
+    httpx_mock.add_response(match_content=b"full content 1")
+    httpx_mock.add_response(match_content=b"full content 2")
+
+    async def stream_content_1() -> AsyncIterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 1"
+
+    async def stream_content_2() -> AsyncIterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 2"
+
+    async with httpx.AsyncClient() as client:
+        await client.put("https://test_url_2", content=stream_content_2())
+        await client.put("https://test_url_1", content=stream_content_1())
+    assert (
+        httpx_mock.get_request(match_content=b"full content 1").url
+        == "https://test_url_1"
+    )
+    assert (
+        httpx_mock.get_request(match_content=b"full content 2").url
+        == "https://test_url_2"
+    )

--- a/tests/test_httpx_sync.py
+++ b/tests/test_httpx_sync.py
@@ -1,4 +1,5 @@
 import re
+from collections.abc import Iterable
 from unittest.mock import ANY
 
 import httpx
@@ -1718,3 +1719,59 @@ def test_custom_transport(httpx_mock: HTTPXMock) -> None:
         response = client.post("https://test_url", content=b"This is the body")
         assert response.read() == b""
         assert response.headers["x-prefix"] == "test"
+
+
+def test_response_selection_content_matching_with_iterable(
+    httpx_mock: HTTPXMock,
+) -> None:
+    httpx_mock.add_response(match_content=b"full content 1", content=b"matched 1")
+    httpx_mock.add_response(match_content=b"full content 2", content=b"matched 2")
+
+    def stream_content_1() -> Iterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 1"
+
+    def stream_content_2() -> Iterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 2"
+
+    with httpx.Client() as client:
+        response_2 = client.put("https://test_url", content=stream_content_2())
+        response_1 = client.put("https://test_url", content=stream_content_1())
+    assert response_1.content == b"matched 1"
+    assert response_2.content == b"matched 2"
+
+
+def test_request_selection_content_matching_with_iterable(
+    httpx_mock: HTTPXMock,
+) -> None:
+    httpx_mock.add_response(match_content=b"full content 1")
+    httpx_mock.add_response(match_content=b"full content 2")
+
+    def stream_content_1() -> Iterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 1"
+
+    def stream_content_2() -> Iterable[bytes]:
+        yield b"full"
+        yield b" "
+        yield b"content"
+        yield b" 2"
+
+    with httpx.Client() as client:
+        client.put("https://test_url_2", content=stream_content_2())
+        client.put("https://test_url_1", content=stream_content_1())
+    assert (
+        httpx_mock.get_request(match_content=b"full content 1").url
+        == "https://test_url_1"
+    )
+    assert (
+        httpx_mock.get_request(match_content=b"full content 2").url
+        == "https://test_url_2"
+    )

